drivers: sensor: Add sensor driver for BH1790

MagpieEmbedded opened the pull request:
This PR adds a sensor driver for the BH1790 Heart Rate Sensor IC ([Datasheet](https://fscdn.rohm.com/en/products/databook/datasheet/ic/sensor/pulse_wave/bh1790glc-e.pdf)).

I have used the BH1750 implementation as a starting point for the build setup and other sensors for implementation. Not all registers and configuration has been implemented. That can be done as part of this PR or as a follow up PR.

A sample has been added which is combined with the MAX30101 heart rate sensor sample. I have separated this into a second commit for the time being so that it can be split into another PR if necessary.

I have tried to follow the contribution guidelines and code style guides, but might have missed some things in the meantime.

Happy to make changes as guided.

Thanks,

Tim
Magpie Embedded

MagpieEmbedded committed:
samples: sensor: Modify MAX30101 sample to generic heart rate sensor sample

This commit modifies the MAX30101 sample to produce a generic heart rate
sensor sample.
Support for the BH1790 sensor is added.

Signed-off-by: Magpie Embedded <magpieembedded@gmail.com>

yperess commented on drivers/sensor/rohm/bh1790/bh1790.c:
```diff
@@ -0,0 +1,171 @@
+	uint16_t led_off_data = 0;
+	uint16_t led_on_data = 0;
+
+	uint8_t value_read_start_address = 0x54;
```
nit: `const`

MagpieEmbedded commented on drivers/sensor/rohm/bh1790/bh1790.c:
```diff
@@ -0,0 +1,171 @@
+	struct bh1790_data *drv_data = dev->data;
+
+	switch (chan) {
+	case SENSOR_CHAN_GREEN:
+		val->val1 = drv_data->led_on_data;
+		val->val2 = 0;
+		break;
+	case SENSOR_CHAN_LIGHT:
+		val->val1 = drv_data->led_off_data;
+		val->val2 = 0;
+		break;
```
Hi @yperess, thanks for reviewing. I have made fixes to the code as you requested above.

Regarding the channels, I chose this to match the other heart rate sensor currently in Zephyr, the MAX30101, in returning data on the Green light channel. The BH1790 also provides "OFF" data which is measuring ambient light. Neither channel on the BH1790 is associated directly with lumens according to the datasheet but this seemed like the best place to put it.

The sensor only returns light readings, it does not return heart rate data directly. That requires further data analysis at the application level.

This seemed like the best compromise to me based on what is already in Zephyr. I would be happy to make further changes to implement this in a better way.

Thanks,

Tim

MagpieEmbedded replied:
Hi @yperess, thanks for your initial review. Pinging to ask if there are any further changes which need to be made.

According to the [MAX30101 datasheet](https://www.analog.com/media/en/technical-documentation/data-sheets/max30101.pdf) (Page 3) and the [BH1790 datasheet](https://fscdn.rohm.com/en/products/databook/datasheet/ic/sensor/pulse_wave/bh1790glc-e.pdf) (Page 6), both sensors return "Counts", not lumens. I am open to adding an additional sensor channel for this. I chose not to make this change in the PR originally because I thought that might be overreaching.

With respect to heart rate measurements, I think that needs to be left in the application layer rather than put into the driver as there are so many different ways of doing this, each with their own tradeoffs.

Thanks,

Tim

MagpieEmbedded replied:
Hi @yperess, I would be happy to change the channel names and also make this change for the MAX30101 sensor so that they are aligned. Just want to make sure this is necessary before I go digging into it. Thanks, Tim

teburd commented on samples/sensor/heart_rate/src/main.c:
```diff
@@ -11,14 +11,14 @@
 
 	if (dev == NULL) {
-		printf("Could not get max30101 device\n");
+		printf("Could not get %s\n", dev->name);
```
This is wrong

MagpieEmbedded replied:
Fixed now 👍

kartben commented on tests/drivers/build_all/sensor/i2c.dtsi:
```diff
@@ -647,6 +647,11 @@ test_i2c_bh1750: bh1750@5e {
 	reg = <0x5e>;
 };
 
+test_i2c_bh1790: bh1790@5b {
+	compatible = "rohm,bh1790";
+	reg = <0x5b>;
+};
+
```
you need to add at the bottom of the list and increment whatever is the last used register by 1. Here you're using 5b which is already taken by maxim,max31875 (and yeah, this means that the PR is then prone to merge conflicts, unfortunately :|)

MagpieEmbedded replied:
Thanks @kartben, I didn't realise that as there were no failure on the test pipelines. Should be fixed now 👍

MagpieEmbedded committed:
drivers: sensor: Add sensor driver for BH1790

This commit adds a driver for the BH1790 Heart Rate Monitor IC.

Based on the approach used by the MAX30101 driver, an existing optical
heart rate sensor with a sample in Zephyr.

Signed-off-by: Magpie Embedded <magpieembedded@gmail.com>

MaureenHelm commented on samples/sensor/heart_rate/boards/nrf52840dk_nrf52840.overlay:
```diff
@@ -0,0 +1,18 @@
+/ {
```
Add license/copyright header

MagpieEmbedded commented:
@MaureenHelm Thanks for your comments, these should be addressed now 👍

MagpieEmbedded commented:
@MaureenHelm Thanks for your approval, is there anything else I can do to help this PR along? It looks like the merge is blocked on the review from @yperess ? Thanks, Tim

kartben commented:
@yperess please revisit

kartben commented:
rebased to address merge conflict in build_all/sensor/i2c.dtsi.
@MaureenHelm please refresh +1 if you don't mind, this has been dragging for way too long IMO :) Thanks!

MagpieEmbedded commented:
Thanks for all the help @kartben, @MaureenHelm and @yperess!
